src/types.ts

```diff
@@ -105,7 +109,7 @@ export type Mapped<T> = Identity<{ [k in keyof T]: T[k] }>
 // Note> Most of the time TNeedsProps is empty, because the overloads in `Connect`
 // just pass in `{}`.  The real props we need come from the component.
 export type InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> = <
-  C extends ComponentType<Matching<TInjectedProps, GetProps<C>>>
+  C extends JSXElementConstructor<Matching<TInjectedProps, GetProps<C>>>
```

Comment:
I'm really not sure I understand why this is OK to swap out for `ComponentType`. Things like [`React.memo` don't use this type](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/v16/index.d.ts#L860), even in the v16 types. It also seems like a fundamentally different type, describing just the construction of an element, not a complete description of a component for other use cases. For instance, things like `displayName` will be missing from this type. I could be misinterpreting, but I don't think this is the correct type to use here.

Reply:
`InferableComponentEnhancerWithProps` is used indirectly to infer props, so we don't need the additional static properties. At least not for the usage with `connect`.

If you look at [React.ComponentProps](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6c3b65cd36ea33204894086cd791ea3adb74f4b/types/react/v16/index.d.ts#L829), you'll see it assumes the type extends either an intrinsic element (which never makes sense to connect anyway) or a JSXElementConstructor.

The issue with [ComponentType](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6c3b65cd36ea33204894086cd791ea3adb74f4b/types/react/v16/index.d.ts#L78) is that it is a union of the class type [`ComponentClass`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6c3b65cd36ea33204894086cd791ea3adb74f4b/types/react/v16/index.d.ts#L583) which [`Component`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d6c3b65cd36ea33204894086cd791ea3adb74f4b/types/react/v16/index.d.ts#L431) (or PureComponent) does not extend.

Ultimately the problem I was encountering is that for a simple PureComponent, `GetProps` was always returning `never`, because the PureComponent class does not extend ComponentType.

Reply:
Could you give a code example that shows how the current types break for v16 and would work for v17+? That may be more helpful for me to understand.

Reply:
Yes, I can add some test coverage to this PR as well. Will do first thing Monday.